### PR TITLE
Makev1

### PR DESCRIPTION
**Pull request by jacobcrawford**
The functionality is not final.
Feedback on the general structure is appreciated.

Questions:
* How should we document targets with `make help`?
* What should the structure of the makefile follow?

### CODE REVIEW

File: config [hunk 1]
```diff
@@ -0,0 +1,3 @@
+NODE_CONTROL_PLANE_VIP=192.168.1.50
+NODE_CONTROL_PLANE_PORT=4200
+NODE_JOIN_TOKEN=1234sdf
```
Comment (skovbakke) [inline config:3]:
```suggestion
NODE_JOIN_TOKEN=1234sdf

```
Add newline at EOF

File: scripts/prepare_master_HA.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+export INTERFACE=$MASTER_NETWORK_INTERFACE
+export ROUTER_ID=51 # Default value
+export PRIORITY=$MASTER_PROXY_PRIORITY
+export AUTH_PASS=42
+
+# Fill in keepalived.conf
+eval "cat <<EOF
+$(<$TEMPLATES_DIR/keepalived.conf)
+EOF
+" > $WORKING_DIR/etc/keepalived/keepalived.conf
```
Comment (skovbakke) [inline scripts/prepare_master_HA.sh:58]:
```suggestion
" > $WORKING_DIR/etc/keepalived/keepalived.conf

```
Newline at EOF

File: scripts/prepare_node_config.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+if [ -z $NODE_TYPE ]
+then
+    NODE_TYPE=worker
+fi
+
+echo "NODE_TYPE=$NODE_TYPE" > $CONFIG_OUTPUT_FILE
+echo "NODE_CONTROL_PLANE_VIP=$NODE_CONTROL_PLANE_VIP" >> $CONFIG_OUTPUT_FILE
+echo "NODE_CONTROL_PLANE_PORT=$NODE_CONTROL_PLANE_PORT" >> $CONFIG_OUTPUT_FILE
+echo "NODE_JOIN_TOKEN=$NODE_JOIN_TOKEN" >> $CONFIG_OUTPUT_FILE
```
Comment (skovbakke) [inline scripts/prepare_node_config.sh:30]:
```suggestion
echo "NODE_JOIN_TOKEN=$NODE_JOIN_TOKEN" >> $CONFIG_OUTPUT_FILE

```
Newline at EOF

File: templates/check_apiserver.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+errorExit() {
+    echo "*** Error happened in check_apiserver.sh" 1>&2
+    exit 1
+}
+
+curl --silent --max-time 2 --insecure https://localhost:${APISERVER_DEST_PORT}/ -o /dev/null || errorExit "Error GET https://localhost:${APISERVER_DEST_PORT}/"
+if ip addr | grep -q ${APISERVER_VIP}; then
+    curl --silent --max-time 2 --insecure https://${APISERVER_VIP}:${APISERVER_DEST_PORT}/ -o /dev/null || errorExit "Error GET https://${APISERVER_VIP}:${APISERVER_DEST_PORT}/"
+fi
```
Comment (skovbakke) [inline templates/check_apiserver.sh:11]:
```suggestion
fi

```
Newline at EOF

File: templates/haproxy.cfg [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+#---------------------------------------------------------------------
+# round robin balancing for apiserver
+#---------------------------------------------------------------------
+backend apiserver
+    option httpchk GET /healthz
+    http-check expect status 200
+    mode tcp
+    option ssl-hello-chk
+    balance     roundrobin
```
Comment (skovbakke) [inline templates/haproxy.cfg:48]:
```suggestion
    balance     roundrobin
    
```
Newline at EOF

File: templates/haproxy.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    - mountPath: /usr/local/etc/haproxy/haproxy.cfg
+      name: haproxyconf
+      readOnly: true
+  hostNetwork: true
+  volumes:
+  - hostPath:
+      path: /etc/haproxy/haproxy.cfg
+      type: FileOrCreate
+    name: haproxyconf
+status: {}
```
Comment (skovbakke) [inline templates/haproxy.yaml:27]:
```suggestion
status: {}

```
Newline at EOF

File: templates/keepalived.conf [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        auth_type PASS
+        auth_pass ${AUTH_PASS}
+    }
+    virtual_ipaddress {
+        ${APISERVER_VIP}
+    }
+    track_script {
+        check_apiserver
+    }
+}
```
Comment (skovbakke) [inline templates/keepalived.conf:29]:
```suggestion
}

```
Newline at EOF

File: templates/keepalived.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      name: check
+  hostNetwork: true
+  volumes:
+  - hostPath:
+      path: /etc/keepalived/keepalived.conf
+    name: config
+  - hostPath:
+      path: /etc/keepalived/check_apiserver.sh
+    name: check
+status: {}
```
Comment (skovbakke) [inline templates/keepalived.yaml:31]:
```suggestion
status: {}

```
Newline at EOF

File: templates/check_apiserver.sh [hunk 9]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+errorExit() {
+    echo "*** Error happened in check_apiserver.sh" 1>&2
+    exit 1
+}
```
Comment (skovbakke) [inline templates/check_apiserver.sh:6]: This does not seem to be using the `arg` you pass to it below, i.e. when you call `errorExit "Error GET https://localhost:${APISERVER_DEST_PORT}/"`.  Is this intentional?

**Review by skovbakke — COMMENTED** [review]
Just some initial comments.

File: scripts/pack_container_images.sh [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+DIR=$1
+mkdir $DIR -p
+i=1
+while read p; do
+  docker pull "$p"
+  docker save --output="$DIR/$i.tar" "$p"
+  i=$((i+1))
+done <requirements.txt
```
Comment (jacobcrawford) [inline scripts/pack_container_images.sh:9]: Check that folder has as many file as there are lines in the requirements.txt file.
Else delete the folder and exit with error.

File: scripts/prepare_cluster.sh [hunk 11]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+OUTPUT_DIR=$1
+VARIABLES=$2
+source $VARIABLES
+
+#TODO validate with regexp
+if [ -z $MASTER_CERTIFICATE_KEY ]
+then
+    echo "[info] MASTER_CERTIFICATE_KEY not set. Generating new."
+    MASTER_CERTIFICATE_KEY=$(docker run kubeadocker certs certificate-key)
```
Comment (jacobcrawford) [inline scripts/prepare_cluster.sh:10]: unknown 'certs' as command to kubeadm

File: config (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacobcrawford) [reply config:3]: Old file. It will be deleted!

File: templates/check_apiserver.sh (hunk 9 above)
Comment (jacobcrawford) [reply templates/check_apiserver.sh]: This has been fixed"

File: templates/boot.sh [hunk 12]
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash 
+sudo swapoff -a
```
Comment (MartinSchmidt) [inline templates/boot.sh:2]: Swap can be permanently disabled in the machines, this only disables it when the script runs, but a reboot would reenable it. 
https://serverfault.com/questions/684771/best-way-to-disable-swap-in-linux
I would be best to make sure Andreas has permanently disabled it, and possibly also `/etc/fstab`

**Review by MartinSchmidt — APPROVED** [review]
I think you should create and issue/story for optimizing the swapoff. see comment

File: templates/boot.sh (hunk 12 above)
Comment (skovbakke) [reply templates/boot.sh:2]: @MartinSchmidt Jeg ville tilføje din kommentar som et issue på [mukube](https://github.com/distributed-technologies/mukube), så vi ikke glemmer det. Men på mukube repositoriet kan man ikke lave issues. Jeg tror det er noget du ikke har slået til eller har slået fra på et tidspunkt.

**Pull request merged**